docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,182 @@
+
+With this implementation the cluster can provide the entire original ipfs file dag on request.  For example if an ipfs peer queries for the entire file they first resolve the root node of the dag which must be pinned on some shard.  If the root has child links then ipfs peers in the cluster are guaranteed to resolve them with content, as all blocks are pinned somewhere in the cluster as children of a shard-node in the cluster-dag.  Note that this implementation conforms well to the goal of coordinating "lightly" above because the cluster state need only keep track of 1 cid per shard.  This prevents the state size from growing too large or more complex than necessary for the purpose of allocation.
+
+The state format would change slightly to account for linking together the root cid of an ipfs file dag and the cluster-dag pinning its leaves
+
+```
+"<cid>": {
+  "name": <pinname>,
+  "cid": <cid>,
+  "clusterdag": <cidcluster>,
+  "allocations" : []
+}
+"<shard1-cid>": {
+  "name": "<pinname>-shard1",
+  "cid": <shard1-cid>,
+  "clusterdag": nil,
+  "allocations": [cluster peers]
+}
+....
+```
+
+Under this implementation replication of shards proceeds like replication of an ordinary pinned cid.  As a consequence shards can be pinned with a replication factor.
+
+### User APIs
+
+#### Importing files for sharding
+Data sharding must begin with a user facing endpoint.  Ideally cluster could stream files, or other data to be imported, across its HTTP API.  This way `ipfs-cluster-ctl` could launch the command for adding and sharding data in the usual way, and users also have the option of hitting the HTTP endpoint over the network.  Multiple files are typically packaged in an HTTP request as the parts of a `Content-Type: multi-part` request and HTTP achieves streaming with `Transfer-Type: chunked`.  Because golang http and mime/multipart [support chunked (a.k.a. streamed) transfer of multipart messages](https://gist.github.com/ZenGround0/49e4a1aa126736f966a1dfdcb84abdae), building the user endpoint into the HTTP API should be relatively straightforward.
+
+The golang multipart-reader exposes a reader to each part of the body in succession.  When the part's body is read completely, a reader to the next part's body can be generated by the multipart-reader.  The part meta-data (e.g. this part is a directory, that part is a file) and the data read from the part, must then be passed into the importer component to begin the next phase of the pipeline.  We can explore different methods of this transfer in pursuit of efficiency and simplicity.  One approach would be polling the part readers in the HTTP API and transferring data as it arrives over RPC, while another would be to assume the importer is on the same machine as the HTTP API and send the multi-part reader, or the readers to individual parts, over a channel.
```

Comment:
I've been thinking a bit, we'll just have an importer that we can instantiate inside the http api. It would be an external library, so the http api component will do the chunking for api uploads. Once a dag block is produced, we send that via RPC to the sharding component.